skills/types-check/SKILL.md

```diff
@@ -0,0 +1,279 @@
+
+## When to Validate Inputs
+
+**Validate at entry points, not everywhere.**
+
+Input validation should happen at the boundary between user code and your package's internal implementation, typically in exported functions that accept user data. Once inputs are validated at these entry points, internal helper functions can trust the data they receive without checking again.
+
+A good analogy to keep in mind is gradual typing. Think of input validation like TypeScript type guards. Once you've validated data at the boundary, you can treat it as "typed" within your internal functions. Additional runtime checks are not needed. The entry point validates once, and all downstream code benefits.
+
+Exception: Validate when in doubt. Do validate in internal functions if:
+- The cost of invalid data is high (data corruption, security issues)
+- The function or context is complex and you want defensive checks
+
+### Entry points (validate here)
+
+- **Exported functions**: Functions users call directly
+- **Functions accepting user data**: Even internal functions if they directly consume user input, or external data (e.g. unserialised data)
+
+```r
+# Exported function: VALIDATE
+#' @export
+create_report <- function(title, n_rows) {
+  check_string(title)
+  check_number_whole(n_rows, min = 1)
+
+  # Now call helpers with validated data
+  data <- generate_data(n_rows)
+  format_report(title, data)
+}
+```
+
+### Internal helpers (don't validate)
+
+Once data is validated at the entry point, internal helpers can skip validation:
+
+```r
+# Internal helper: NO VALIDATION NEEDED
+generate_data <- function(n_rows) {
+  # n_rows is already validated, just use it
+  data.frame(
+    id = seq_len(n_rows),
+    value = rnorm(n_rows)
+  )
+}
+
+# Internal helper: NO VALIDATION NEEDED
+format_report <- function(title, data) {
+  # title and data are already validated, just use them
+  list(
+    title = title,
+    summary = summary(data),
+    rows = nrow(data)
+  )
+}
+```
+
+Note how the `data` generated by `generate_data()` doesn't need validation either. Internal code creating data in a trusted way (e.g. because it's simple or because it's covered by unit tests) doesn't require internal checks.
+
+## Early input checking
+
+Always validate inputs at the start of user-facing functions, before doing any work:
+
+```r
+my_function <- function(x, name, env = caller_env()) {
+  check_logical(x)
+  check_name(name)
+  check_environment(env)
+
+  # ... function body
+}
+```
+
+Benefits:
+
+- This self-documents the types of the arguments
+- Eager evaluation also reduces the risk of confusing lazy evaluation effects
+
+## Choosing the Right Checker
+
+### Scalars (single values)
+
+For atomic vectors, use scalar checkers when arguments parameterise the function (configuration flags, names, single counts), rather than represent vectors of user data. They assert a single value.
+
+- `check_bool()`: Single TRUE/FALSE (use for flags/options)
```

Comment:
I'd move these up to the top so there's a concrete list of checker early on in the doc.